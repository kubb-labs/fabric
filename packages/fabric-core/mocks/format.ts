import type { Options } from 'prettier'
import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }

  try {
    return prettierFormat(source, formatOptions)
  } catch (_e) {
    return Promise.resolve(source)
  }
}
