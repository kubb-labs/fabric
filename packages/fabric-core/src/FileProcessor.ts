import type * as KubbFile from './KubbFile.ts'
import { write } from './fs.ts'
import pLimit from 'p-limit'
import type { Parser } from './parsers/types.ts'
import { defaultParser } from './parsers/default.ts'
import { AsyncEventEmitter } from './utils/AsyncEventEmitter.ts'
import type { Events } from './defineApp.ts'
import { typeScriptParser } from './parsers/typescript.ts'
import { tsxParser } from './parsers/tsx.ts'

export type ProcessFilesProps = {
  parsers?: Record<KubbFile.Extname, Parser<any>>
  extension?: Record<KubbFile.Extname, KubbFile.Extname | ''>
  dryRun?: boolean
}

type GetParseOptions = {
  extname?: KubbFile.Extname
  parsers?: Record<KubbFile.Extname, Parser<any>>
}

type Options = {
  events?: AsyncEventEmitter<Events>
}

export class FileProcessor {
  #limit = pLimit(100)
  events: AsyncEventEmitter<Events>

  constructor({ events = new AsyncEventEmitter<Events>() }: Options = {}) {
    this.events = events

    return this
  }

  get defaultParser(): Record<KubbFile.Extname, Parser<any>> {
    return {
      '.ts': typeScriptParser,
      '.js': typeScriptParser,
      '.jsx': tsxParser,
      '.tsx': tsxParser,
      '.json': defaultParser,
    }
  }

  async parse(file: KubbFile.ResolvedFile, { parsers = this.defaultParser, extname }: GetParseOptions = {}): Promise<string> {
    if (!extname) {
      console.warn(`[parser] No parser found for ${extname}, default parser will be used`)
      return defaultParser.print(file, { extname })
    }

    const parser = parsers[extname]

    if (!parser) {
      console.warn(`[parser] No parser found for ${extname}, default parser will be used`)

      return defaultParser.print(file, { extname })
    }

    return parser.print(file, { extname })
  }

  async run(files: Array<KubbFile.ResolvedFile>, { parsers, dryRun, extension }: ProcessFilesProps = {}): Promise<KubbFile.ResolvedFile[]> {
    await this.events.emit('process:start', { files })

    let processed = 0
    const total = files.length

    const promises = files.map((resolvedFile, index) =>
      this.#limit(async () => {
        const extname = extension?.[resolvedFile.extname] || undefined

        await this.events.emit('file:start', { file: resolvedFile, index, total })

        await this.events.emit('file:start', { file: resolvedFile, index, total })

        if (!dryRun) {
          const source = await this.parse(resolvedFile, { extname, parsers })
          await write(resolvedFile.path, source, { sanity: false })
        }

        await this.events.emit('process:progress', { file: resolvedFile, processed, percentage: (processed / total) * 100, total })

        await this.events.emit('file:end', { file: resolvedFile, index, total })

        processed++
      }),
    )

    await Promise.all(promises)

    await this.events.emit('process:end', { files })

    return files
  }
}
