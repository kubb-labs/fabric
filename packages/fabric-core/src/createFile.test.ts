import { format } from '../mocks/format.ts'
import { combineImports, combineSources, createFile, combineExports } from './createFile.ts'

import type * as KubbFile from './KubbFile.ts'
import { parseFile } from './FileProcessor.ts'

describe('createFile', () => {
  test('if getFileSource is returning code with imports', async () => {
    const code = await parseFile(
      createFile({
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        imports: [
          {
            name: ['Pets'],
            path: './Pets.ts',
            isTypeOnly: true,
          },
        ],
        sources: [
          {
            value: 'export type Pet = Pets;',
          },
        ],
      }),
    )

    const codeWithDefaultImport = await parseFile(
      createFile({
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        imports: [
          {
            name: 'client',
            path: './Pets.ts',
          },
          {
            name: ['Pets', 'Cat'],
            path: './Pets.ts',
            isTypeOnly: true,
          },
          {
            name: 'React',
            path: './React.ts',
          },
        ],
        sources: [
          {
            value: 'export type Pet = Pets | Cat; const test = [client, React];',
          },
        ],
      }),
    )

    const codeWithDefaultImportOrder = await parseFile(
      createFile({
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        sources: [
          {
            value: 'export type Pet = Pets | Cat;\nconst test = [client, React];',
          },
        ],
        imports: [
          {
            name: ['Pets', 'Cat'],
            path: './Pets.ts',
            isTypeOnly: true,
          },
          {
            name: 'client',
            path: './Pets.ts',
          },
          {
            name: 'React',
            path: './React.ts',
          },
          {
            name: ['Pets', 'Cat'],
            path: './Pets.ts',
            isTypeOnly: true,
          },
        ],
      }),
    )

    expect(await format(code)).toMatchSnapshot()
    expect(await format(codeWithDefaultImport)).toMatchSnapshot()
    expect(await format(codeWithDefaultImportOrder)).toMatchSnapshot()
  })

  test('if getFileSource is returning code with imports and default import', async () => {
    const code = await parseFile(
      createFile({
        baseName: 'test.ts',
        path: 'models/ts/test.ts',
        sources: [
          {
            value: 'export type Pet = Pets;',
          },
        ],
        imports: [
          {
            name: 'Pets',
            path: './Pets.ts',
            isTypeOnly: true,
          },
        ],
      }),
    )
    expect(await format(code)).toMatchSnapshot()
  })

  test('if getFileSource is returning code with exports and exports as', async () => {
    const fileImport = createFile({
      path: './src/models/file1.ts',
      baseName: 'file1.ts',
      sources: [
        {
          value: `export const test = 2;
        type Test = Pets | Lily | Dog;`,
        },
      ],
      imports: [
        {
          name: ['Pets'],
          path: './Pets.ts',
          isTypeOnly: true,
        },
        {
          name: ['Lily'],
          path: './Pets.ts',
          isTypeOnly: true,
        },
        {
          name: 'Dog',
          path: './Dog.ts',
          isTypeOnly: true,
        },
      ],
    })

    const fileExport = createFile({
      path: './src/models/file1.ts',
      baseName: 'file1.ts',
      sources: [],
      exports: [
        {
          name: ['Pets'],
          path: './Pets.ts',
          isTypeOnly: true,
        },
        {
          name: ['Lily'],
          path: './Pets.ts',
          isTypeOnly: true,
        },
        {
          name: 'Dog',
          asAlias: true,
          path: './Dog.ts',
          isTypeOnly: true,
        },
      ],
    })

    expect(await format(await parseFile(fileImport))).toMatchSnapshot()
    expect(await format(await parseFile(fileExport))).toMatchSnapshot()
  })

  test('if combineExports is filtering out duplicated sources(by name)', () => {
    const sources: Array<KubbFile.Source> = [
      {
        name: 'test',
        isTypeOnly: false,
        value: 'const test = 2',
      },
      {
        name: 'test',
        isTypeOnly: false,
        value: 'const test = 3',
      },
      {
        name: 'Test',
        isTypeOnly: false,
        value: 'type Test = 2',
      },
    ]

    expect(combineSources(sources)).toMatchInlineSnapshot(`
      [
        {
          "isTypeOnly": false,
          "name": "test",
          "value": "const test = 2",
        },
        {
          "isTypeOnly": false,
          "name": "test",
          "value": "const test = 3",
        },
        {
          "isTypeOnly": false,
          "name": "Test",
          "value": "type Test = 2",
        },
      ]
    `)
  })

  test('if combineExports is filtering out duplicated exports(by path and name)', () => {
    const exports: Array<KubbFile.Export> = [
      {
        path: './models',
        name: undefined,
        isTypeOnly: true,
      },
      {
        path: './models',
        isTypeOnly: false,
      },
      {
        path: './models',
        isTypeOnly: false,
        asAlias: true,
        name: 'test',
      },
    ]

    expect(combineExports(exports)).toMatchInlineSnapshot(`
      [
        {
          "isTypeOnly": true,
          "name": undefined,
          "path": "./models",
        },
        {
          "asAlias": true,
          "isTypeOnly": false,
          "name": "test",
          "path": "./models",
        },
      ]
    `)
  })

  test('if combineImports is filtering out duplicated imports(by path and name)', () => {
    const imports: Array<KubbFile.Import> = [
      {
        path: './models',
        name: 'models',
        isTypeOnly: true,
      },
      {
        path: './models',
        name: ['Config'],
        isTypeOnly: true,
      },
      {
        path: './models',
        name: 'models',
        isTypeOnly: false,
      },
    ]

    expect(combineImports(imports, [], 'const test = models; type Test = Config;')).toMatchInlineSnapshot(`
      [
        {
          "isTypeOnly": true,
          "name": "models",
          "path": "./models",
        },
        {
          "isTypeOnly": true,
          "name": [
            "Config",
          ],
          "path": "./models",
        },
      ]
    `)

    const importsWithoutSource: Array<KubbFile.Import> = [
      {
        path: './models',
        name: 'models',
        isTypeOnly: true,
      },
      {
        path: './models',
        name: ['Config'],
        isTypeOnly: true,
      },
      {
        path: './models',
        name: 'models',
        isTypeOnly: false,
      },
    ]

    expect(combineImports(importsWithoutSource, [])).toEqual([imports[0], imports[1]])
  })
})
