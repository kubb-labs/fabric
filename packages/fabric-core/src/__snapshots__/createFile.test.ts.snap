// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createFile > if getFileSource is returning code with exports and exports as 1`] = `
"import type Dog from './Dog.ts'
import type { Pets, Lily } from './Pets.ts'

export const test = 2
type Test = Pets | Lily | Dog
"
`;

exports[`createFile > if getFileSource is returning code with exports and exports as 2`] = `
"export type * as Dog from './Dog.ts'
export type { Pets, Lily } from './Pets.ts'
"
`;

exports[`createFile > if getFileSource is returning code with imports 1`] = `
"import type { Pets } from './Pets.ts'

export type Pet = Pets
"
`;

exports[`createFile > if getFileSource is returning code with imports 2`] = `
"export type Pet = Pets | Cat
const test = [client, React]
"
`;

exports[`createFile > if getFileSource is returning code with imports 3`] = `
"import client from './Pets.ts'
import React from './React.ts'
import type { Pets, Cat } from './Pets.ts'

export type Pet = Pets | Cat
const test = [client, React]
"
`;

exports[`createFile > if getFileSource is returning code with imports and default import 1`] = `
"import type Pets from './Pets.ts'

export type Pet = Pets
"
`;
